File: src/main/java/ninja/leaping/permissionsex/sponge/PEXSpongeCommand.java
/**
 * PermissionsEx
 * Copyright (C) zml and PermissionsEx contributors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package ninja.leaping.permissionsex.sponge;

import ninja.leaping.permissionsex.util.command.CommandSpec;
import org.spongepowered.api.text.Text;
import org.spongepowered.api.util.command.CommandCallable;
import org.spongepowered.api.util.command.CommandException;
import org.spongepowered.api.util.command.CommandSource;

import java.util.List;

/**
 * Wrapper class between PEX commands and the Sponge command class
 */
public class PEXSpongeCommand implements CommandCallable {
    private final CommandSpec command;
    private final PermissionsExPlugin plugin;

    public PEXSpongeCommand(CommandSpec command, PermissionsExPlugin plugin) {
        this.command = command;
        this.plugin = plugin;
    }

    @Override
    public boolean call(CommandSource commandSource, String arguments, List<String> parents) throws CommandException {
        command.process(new SpongeCommander(plugin, commandSource), arguments);
        return true;
    }

    @Override
    public boolean testPermission(CommandSource commandSource) {
        try {
            command.checkPermission(new SpongeCommander(plugin, commandSource));
        } catch (ninja.leaping.permissionsex.util.command.CommandException e) {
            return false;
        }
        return true;
    }

    @Override
    public String getShortDescription(CommandSource commandSource) {
        return null;
    }

    @Override
    public Text getHelp(CommandSource commandSource) {
        return command.getExtendedDescription(new SpongeCommander(plugin, commandSource));
    }

    @Override
    public String getUsage(CommandSource commandSource) {
        return null;
    }

    @Override
    public List<String> getSuggestions(CommandSource commandSource, String commandLine) throws CommandException {
        return command.tabComplete(new SpongeCommander(plugin, commandSource), commandLine);
    }
}

File: platform/sponge7/src/main/kotlin/ca/stellardrift/permissionsex/sponge/PEXSpongeCommand.kt
/*
 * PermissionsEx
 * Copyright (C) zml and PermissionsEx contributors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package ca.stellardrift.permissionsex.sponge

import ca.stellardrift.permissionsex.impl.commands.parse.CommandException
import ca.stellardrift.permissionsex.impl.commands.parse.CommandSpec
import ca.stellardrift.permissionsex.util.optionally
import java.util.Optional
import org.spongepowered.api.command.CommandCallable
import org.spongepowered.api.command.CommandException as SpongeCommandException
import org.spongepowered.api.command.CommandResult
import org.spongepowered.api.command.CommandSource
import org.spongepowered.api.text.Text
import org.spongepowered.api.world.Location
import org.spongepowered.api.world.World

/**
 * Wrapper class between PEX commands and the Sponge command class
 */
internal class PEXSpongeCommand(private val command: CommandSpec, private val plugin: PermissionsExPlugin) : CommandCallable {

    @Throws(SpongeCommandException::class)
    override fun process(commandSource: CommandSource, arguments: String): CommandResult {
        this.command.process(SpongeCommander(this.plugin, commandSource), arguments)
        return CommandResult.success()
    }

    override fun testPermission(commandSource: CommandSource): Boolean {
        try {
            this.command.checkPermission(SpongeCommander(this.plugin, commandSource))
        } catch (e: CommandException) {
            return false
        }
        return true
    }

    override fun getShortDescription(commandSource: CommandSource): Optional<Text> {
        return this.command.description?.toSponge().optionally()
    }

    override fun getHelp(commandSource: CommandSource): Optional<Text> {
        return this.command.getExtendedDescription(SpongeCommander(plugin, commandSource)).toSponge().optionally()
    }

    override fun getUsage(commandSource: CommandSource): Text {
        return command.getUsage(SpongeCommander(plugin, commandSource)).toSponge()
    }

    @Throws(SpongeCommandException::class)
    override fun getSuggestions(commandSource: CommandSource, commandLine: String, targetPosition: Location<World>?): List<String> {
        return command.tabComplete(SpongeCommander(plugin, commandSource), commandLine)
    }
}
